c++/src/kj/compat/http.c++

```diff
@@ -5716,14 +5716,111 @@ public:
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryGetLength();
+              } else {
+                return unencryptedStream->tryGetLength();
+              }
+            }
+
+            kj::Promise<uint64_t> pumpTo(kj::AsyncOutputStream& output, uint64_t amount) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->pumpTo(output, amount);
+              } else {
+                return unencryptedStream->pumpTo(output, amount);
+              }
+            }
+
+            kj::Promise<void> write(const void* buffer, size_t size) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(buffer, size);
+              } else {
+                return unencryptedStream->write(buffer, size);
+              }
+            }
+
+            kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->write(pieces);
+              } else {
+                return unencryptedStream->write(pieces);
+              }
+            }
+
+            kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(
+                kj::AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->tryPumpFrom(input, amount);
+              } else {
+                return unencryptedStream->tryPumpFrom(input, amount);
+              }
+            }
+
+            kj::Promise<void> whenWriteDisconnected() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->whenWriteDisconnected();
+              } else {
+                return unencryptedStream->whenWriteDisconnected();
+              }
+            }
+
+            void shutdownWrite() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->shutdownWrite();
+              } else {
+              unencryptedStream->shutdownWrite();
+              }
+            }
+
+            void abortRead() override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                s->get()->abortRead();
+              } else {
+                unencryptedStream->abortRead();
+              }
+            }
+
+            kj::Maybe<int> getFd() const override {
+              KJ_IF_MAYBE(s, encryptedStream) {
+                return s->get()->getFd();
+              } else {
+                return unencryptedStream->getFd();
+              }
+            }
+
+            void startTls(
+                kj::SecureNetworkWrapper* wrapper, kj::StringPtr expectedServerHostname) {
+              kj::Promise<kj::Own<kj::AsyncIoStream>> secureStream =
+                  wrapper->wrapClient(kj::mv(unencryptedStream), expectedServerHostname);
+              encryptedStream = kj::newPromisedStream(kj::mv(secureStream));
```

Comment:
If you assigned this into `unencryptedStream` and never touched `encryptedStream`, the code would still work exactly the same.

Maybe you should just have one member `kj::Own<kj::AsyncIoStream> inner` which you overwrite here? Then all the other methods can be simplified to just forward to `inner`, no need to check the `Maybe`.